### PR TITLE
Added typegen blog

### PR DESCRIPTION
**Pull request by mattpocock**

### CODE REVIEW

**Comment by farskid** [comment]
Oh it was already done in https://github.com/statelyai/eng-blog/pull/58 but we can merge it.

**Review by farskid — APPROVED** [review]

**Comment by mattpocock** [comment]
@farskid Whoops, committed the blog to this PR as well

File: content/posts/2022-01-27-introducing-typegen.mdx [hunk 1]
```diff
@@ -0,0 +1,275 @@
+---
+title: "Introducing: TypeScript typegen for XState"
+description: "XState's TypeScript experience just got an enormous upgrade - Typescript typege, integrated with VSCode."
```
Comment (davidkpiano) [inline content/posts/2022-01-27-introducing-typegen.mdx:3]:
```suggestion
description: "XState's TypeScript experience just got an enormous upgrade - Typescript typegen, integrated with VSCode."
```

File: content/posts/2022-01-27-introducing-typegen.mdx [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+tags:
+  - announcement
+  - xstate
+  - community
+  - typescript
+  - tutorial
```
Comment (davidkpiano) [inline content/posts/2022-01-27-introducing-typegen.mdx:9]:
```suggestion
  - tutorial
  - vscode
```

File: content/posts/2022-01-27-introducing-typegen.mdx [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+    },
+  },
+});
+```
+
+3. Add `tsTypes: {}` to the machine and save the file:
+
+```ts
+const machine = createMachine({
+  tsTypes: {},
```
Comment (davidkpiano) [inline content/posts/2022-01-27-introducing-typegen.mdx:64]:
```suggestion
```diff
const machine = createMachine({
+ tsTypes: {},
```

Nit suggestion

File: content/posts/2022-01-27-introducing-typegen.mdx [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+           */
+          console.log(event.value);
+        }
+      },
+    },
+  }
+);
+```
+
+The VSCode extension **statically analyzes your machine**, and knows which events lead to which actions.
```
Comment (davidkpiano) [inline content/posts/2022-01-27-introducing-typegen.mdx]:
```suggestion
The VS Code extension **statically analyzes your machine**, and knows which events lead to which actions.
```

**Review by davidkpiano — APPROVED** [review]
Some copy nits, LGTM otherwise

File: content/posts/2022-01-27-introducing-typegen.mdx [hunk 5]
```diff
@@ -0,0 +1,275 @@
+---
+title: "Introducing: TypeScript typegen for XState"
+description: "XState's TypeScript experience just got an enormous upgrade - Typescript typege, integrated with VS Code."
```
Comment (laurakalbag) [inline content/posts/2022-01-27-introducing-typegen.mdx:3]:
```suggestion
description: "XState’s TypeScript experience just got an enormous upgrade - Typescript typegen, integrated with VS Code."
```

File: content/posts/2022-01-27-introducing-typegen.mdx [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+  - typescript
+  - tutorial
+author:
+  - Matt Pocock
+originalURL: ""
+excerpt: "XState and TypeScript are a match made in heaven, and XState's TypeScript experience just got an enormous upgrade"
+publishedAt: "2022-01-27"
+---
+
+**XState and TypeScript are a match made in heaven**. TypeScript gives you type safety, and XState gives you logical safety. Together, they give you confidence that your code will do what you expect it to.
```
Comment (laurakalbag) [inline content/posts/2022-01-27-introducing-typegen.mdx:17]:
```suggestion
**XState and TypeScript are a match made in heaven**. TypeScript gives you type safety, and XState gives you logical safety. Together, they give you confidence that your code will do what you expect.
```

File: content/posts/2022-01-27-introducing-typegen.mdx [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+
+However, we've been hearing from the community for some time that the _experience_ of using TypeScript with XState needed improving.
+
+Today's your lucky day. **XState's TypeScript experience just got an _enormous_ upgrade**.
+
+We have brought type generation into XState v4.29.0, and we're aiming for **perfect types on all parts of XState's API**.
```
Comment (laurakalbag) [inline content/posts/2022-01-27-introducing-typegen.mdx:23]:
```suggestion
We have brought type generation into XState v4.29.0, and we’re aiming for **perfect types on all parts of XState’s API**.
```

File: content/posts/2022-01-27-introducing-typegen.mdx [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+
+We're releasing it today as an opt-in beta. Be sure to check out the [known limitations](https://xstate.js.org/docs/guides/typescript.html#known-limitations) section of the docs.
```
Comment (laurakalbag) [inline content/posts/2022-01-27-introducing-typegen.mdx:25]:
```suggestion
We’re releasing typegen today as an opt-in beta. Be sure to check out the [known limitations](https://xstate.js.org/docs/guides/typescript.html#known-limitations) section of the docs.
```

File: content/posts/2022-01-27-introducing-typegen.mdx [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+
+## Getting Started
+
+You can visit the [official docs](https://xstate.js.org/docs/guides/typescript.html#typegen-with-the-vscode-extension) for the full guide, or follow the instructions here:
```
Comment (laurakalbag) [inline content/posts/2022-01-27-introducing-typegen.mdx:29]:
```suggestion
You can visit the [official docs](https://xstate.js.org/docs/guides/typescript.html#typegen-with-the-vscode-extension) for the full guide, or follow the instructions:
```

File: content/posts/2022-01-27-introducing-typegen.mdx [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+
+```ts
+const machine = createMachine({
+  // This generic just got added!
+  tsTypes: {} as import("./filename.typegen").Typegen0,
+  /* ... */
+});
+```
+
+5. Add a second parameter into the `createMachine` call - this is where you implement the actions, services, guards and delays for the machine.
```
Comment (laurakalbag) [inline content/posts/2022-01-27-introducing-typegen.mdx:86]: *this* — is the *this* where you implement the actions, services, guards and delays the parameter or the call?

File: content/posts/2022-01-27-introducing-typegen.mdx [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+        console.log(event.value);
+      },
+    },
+  }
+);
+```
+
+## Typing improvements
+
+Let's go into the nitty gritty and show you exactly what's improved.
```
Comment (laurakalbag) [inline content/posts/2022-01-27-introducing-typegen.mdx:110]:
```suggestion
Let’s get into the nitty-gritty and show you exactly what’s improved.
```

File: content/posts/2022-01-27-introducing-typegen.mdx [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+);
+```
+
+This works for actions, services, guards and delays. It even works for **entry actions**, another big pain point from the community.
+
+We're hoping this lets you cut hundreds of lines of useless defensive code.
+
+### Autocomplete on machine options
+
+Another thing we've been hearing from the community is that it's easy to make typos on machine options.
```
Comment (laurakalbag) [inline content/posts/2022-01-27-introducing-typegen.mdx:177]:
```suggestion
Another thing we’ve been hearing from the community is that it’s easy to make typos on machine options.
```

File: content/posts/2022-01-27-introducing-typegen.mdx [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+
+Now, with typegen, you get **autocomplete on all machine options**. This will error:
```
Comment (laurakalbag) [inline content/posts/2022-01-27-introducing-typegen.mdx:179]:
```suggestion
Now, with typegen, you get **autocomplete on all machine options**. The following code will error:
```

File: content/posts/2022-01-27-introducing-typegen.mdx [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+       * This will error, because sayhello does not
+       * exist in the machine declaration above
+       */
+      sayhello: () => {},
+    },
+  }
+);
+```
+
+We've also made it so any _missing_ machine options will error when
```
Comment (laurakalbag) [inline content/posts/2022-01-27-introducing-typegen.mdx:198]:
```suggestion
We’ve also made it so any _missing_ machine options will error when
```

File: content/posts/2022-01-27-introducing-typegen.mdx [hunk 15]
```diff
@@ -0,0 +1,275 @@
...
+const App = () => {
+  /**
+   * This will error, because you haven't implemented
+   * sayHello in your actions object
+   */
+  const [state, send] = useMachine(machine);
+};
+```
+
+That gives you 100% confidence that your machine has all the things it needs to work.
```
Comment (laurakalbag) [inline content/posts/2022-01-27-introducing-typegen.mdx:215]: *That* - is *that* the lack of errors?

File: content/posts/2022-01-27-introducing-typegen.mdx [hunk 16]
```diff
@@ -0,0 +1,275 @@
...
+
+### Typing of promise-services
+
+Using promise-based services might be the single biggest pain point with XState and TypeScript. We used to have an entire troubleshooting section in our docs dedicated to them.
+
+Now, you can **strongly type the results of promise-based services**. Here's how:
```
Comment (laurakalbag) [inline content/posts/2022-01-27-introducing-typegen.mdx:221]:
```suggestion
Now, you can **strongly type the results of promise-based services**. Here’s how:
```

File: content/posts/2022-01-27-introducing-typegen.mdx [hunk 17]
```diff
@@ -0,0 +1,275 @@
...
+         * receive a data attribute with { id: string }
+         */
+        console.log(event.data.id);
+      },
+    },
+  }
+);
+```
+
+This makes handling data return types in XState intuitive, easy and type-safe.
```
Comment (laurakalbag) [inline content/posts/2022-01-27-introducing-typegen.mdx:271]: *this* - is *this* the strongly-typed results?

File: content/posts/2022-01-27-introducing-typegen.mdx [hunk 18]
```diff
@@ -0,0 +1,275 @@
...
+
+## Acknowledgements
+
+I want to thank my Stately colleague [Andarist](https://twitter.com/AndaristRake). His TypeScript wizardry, incredible attention to detail and deep love of open source helped make this possible. We've literally been talking about this for 18 months - and it's finally here.
```
Comment (laurakalbag) [inline content/posts/2022-01-27-introducing-typegen.mdx:275]:
```suggestion
I want to thank my Stately colleague [Andarist](https://twitter.com/AndaristRake). His TypeScript wizardry, incredible attention to detail and deep love of open source helped make this possible. We’ve literally been talking about this for 18 months - and it’s finally here.
```

**Pull request merged**